src/build/source_anchor.rs

```diff
@@ -0,0 +1,75 @@
+use pixi_record::PinnedSourceSpec;
+use pixi_spec::{GitSpec, PathSourceSpec, SourceSpec, UrlSourceSpec};
+use typed_path::Utf8TypedPath;
+
+/// `SourceAnchor` represents the resolved base location of a `SourceSpec`.
+/// It serves as a reference point for interpreting relative or recursive
+/// source specifications, enabling consistent resolution of nested sources.
+pub enum SourceAnchor {
+    /// The source is relative to the workspace root.
+    Workspace,
+
+    /// The source is relative to another source package.
+    Source(SourceSpec),
+}
+
+impl From<PinnedSourceSpec> for SourceAnchor {
+    fn from(value: PinnedSourceSpec) -> Self {
+        SourceAnchor::Source(SourceSpec::from(value))
+    }
+}
+
+impl SourceAnchor {
+    /// Resolve a source spec relative to this anchor.
+    pub fn resolve(&self, spec: SourceSpec) -> SourceSpec {
+        // If this instance is already anchored to the workspace we can simply return
+        // immediately.
+        let SourceAnchor::Source(base) = self else {
+            return match spec {
+                SourceSpec::Url(url) => SourceSpec::Url(url),
+                SourceSpec::Git(git) => SourceSpec::Git(git),
+                SourceSpec::Path(PathSourceSpec { path }) => SourceSpec::Path(PathSourceSpec {
+                    // Normalize the input path.
+                    path: path.normalize(),
+                }),
+            };
+        };
+
+        // Only path specs can be relative.
+        let SourceSpec::Path(PathSourceSpec { path }) = spec else {
+            return spec;
+        };
+
+        // If the path is absolute we can just return it.
+        if path.is_absolute() || path.starts_with("~") {
```

Comment:
Why `~`, are we resolving that somewhere? I expected that we are not able to deal with that.

Reply:
Its part of the `resolve_path` function in BuildContext.